### PR TITLE
CB2-3613: Fix Test Result Data Field Mapping

### PR DESCRIPTION
**Pull request by tom-bjss**
This bug occurs because the wrong field from the API was being used to populate the front end model. The root cause of this was ambiguity in the original ticket which stated to use 'test result status' which does not exist. However, test status and test result both do exist.

A simple change to use the correct api field and update some tests.

Some quality of life changes included:

- trim of the end of the search input
- added backend spin up to the makefile
- updated and extended mock api (thanks @andrewf-bjss)

### CODE REVIEW

File: src/app/features/global-error/global-error.component.html [hunk 1]
```diff
@@ -1,12 +1,10 @@
-<ng-container *ngIf="globalErrorService.errors$ | async as errorMessages">
```
Comment (tom-bjss) [inline src/app/features/global-error/global-error.component.html:1]: Will update this

File: mock-api/mock-api-server.ts [hunk 2]
```diff
@@ -1,19 +1,50 @@
 import * as jsonServer from 'json-server';
 import { mockTestResultList } from '../src/mocks/mock-test-result';
-import { mockVehicleTecnicalRecordList } from '../src/mocks/mock-vehicle-technical-record.mock';
+import { mockVehicleTechnicalRecordList } from '../src/mocks/mock-vehicle-technical-record.mock';
 const server = jsonServer.create();
 const router = jsonServer.router('{}');
 const middlewares = jsonServer.defaults();
 
+const args = require('minimist')(process.argv.slice(2));
```
Comment (kpearmanbjss) [inline mock-api/mock-api-server.ts:8]: Can we have this as an import and add to dev dependencies in package.json?

File: src/app/features/test-record-summary/test-record-summary.component.spec.ts [hunk 3]
```diff
@@ -2,16 +2,35 @@ import { ComponentFixture, TestBed } from '@angular/core/testing';
 import { By } from '@angular/platform-browser';
 import { TestRecordSummaryComponent } from './test-record-summary.component';
 import { TestResultModel } from '../../models/test-result.model';
+import { TestType } from '@models/test-type.model';
 
 const fakeRecord: TestResultModel = {
   testResultId: 'test',
   systemNumber: 'test',
   testStartTimestamp: new Date(),
```
Comment (kpearmanbjss) [inline src/app/features/test-record-summary/test-record-summary.component.spec.ts:10]: If this is set to Date type, do 21 and 28 need to be the same? Currently string

Comment (kpearmanbjss) [inline src/app/features/test-record-summary/test-record-summary.component.spec.ts:10]: If this is set to Date type, do 21 and 28 need to be the same? Currently string

**Review by aaron-downing — APPROVED** [review]

File: src/app/features/test-record-summary/test-record-summary.component.spec.ts (hunk 3 above)
Comment (tom-bjss) [reply src/app/features/test-record-summary/test-record-summary.component.spec.ts:10]: Model accepts string or date type :D

File: README.md [hunk 4]
```diff
@@ -20,6 +20,9 @@ Then to run tests run:
 To just run the application run:
 `make run`
 
+To run the mock api run:
+`make api`
+
```
Comment (andrewf-bjss) [inline README.md:25]: Here's a little more information for the readme about the command line switches added to the mock-api:

```suggestion
To run the mock api run:
`make api`

The mock api also accepts optional command line switches to affect the response:
`npm run mock-api -- --tech-record=[option] --test-result=[option]`

where `[option]` is:
- `NotFound` - returns a 404
- `ServiceError` - returns a 500
```

File: src/app/features/test-record-summary/test-record-summary.component.spec.ts (hunk 3 above)
Comment (andrewf-bjss) [reply src/app/features/test-record-summary/test-record-summary.component.spec.ts:10]: I think this import might need to be:
`import { TestType } from '../../models/test-type.model';`

File: mock-api/mock-api-server.ts (hunk 2 above)
Comment (tom-bjss) [reply mock-api/mock-api-server.ts:8]: From documentation https://www.npmjs.com/package/minimist
Looks like this is the desired pattern

File: README.md (hunk 4 above)
Comment (tom-bjss) [reply README.md]: Updated with a nice table

File: src/app/features/test-record-summary/test-record-summary.component.spec.ts (hunk 3 above)
Comment (tom-bjss) [reply src/app/features/test-record-summary/test-record-summary.component.spec.ts:10]: Updated

File: README.md [hunk 5]
```diff
@@ -20,6 +20,19 @@ Then to run tests run:
...
+| Argument                      | Tech Record API | Test Record API |
+|-------------------------------|-----------------|-----------------|
+| -- --tech-record=NotFound     | 404             | -               |
+| -- --tech-record=ServiceError | 500             | -               |
+| -- --test-result=NotFound     | 200             | 404             |
+| -- --test-result=ServiceError | 200             | 500             |
+
+
+
```
Comment (andrewf-bjss) [inline README.md:35]: How about this tweak?

```suggestion
To run the mock api run:
`make api`
or use `npm run mock-api` if you need to pass arguments to affect the HTTP response:

| Argument                        | Tech Record API Response | Test Result API Response |
|---------------------------------|--------------------------|--------------------------|
| `-- --tech-record=NotFound`     | 404                      | -                        |
| `-- --tech-record=ServiceError` | 500                      | -                        |
| `-- --test-result=NotFound`     | 200                      | 404                      |
| `-- --test-result=ServiceError` | 200                      | 500                      |

```

File: src/app/features/search/search.component.ts [hunk 6]
```diff
@@ -23,7 +23,7 @@ export class SearchComponent {
     this.globalErrorService.errors = [];
     this.searchError = undefined;
     if (searchTerm) {
-      this.technicalRecordService.searchBy({ type: 'vin', searchTerm });
+      this.technicalRecordService.searchBy({ type: 'vin', searchTerm: searchTerm.trim() });
```
Comment (mcadenas-bjss) [inline src/app/features/search/search.component.ts:26]: `trim()` should be done on or before checking value of `searchTerm` as `'    '` would still evaluate to `true` and send an empty string to `searchBy` method.

File: src/app/features/test-record-summary/test-record-summary.component.spec.ts (hunk 3 above)
Comment (mcadenas-bjss) [reply src/app/features/test-record-summary/test-record-summary.component.spec.ts:10]: `@models/` is an alias for `models` folder so both are valid. you can see path aliases defined in `tsconfig.json` e.g. `"@models/*": ["src/app/models/*"],`

File: src/app/features/test-record-summary/test-record-summary.component.spec.ts [hunk 7]
```diff
@@ -2,16 +2,35 @@ import { ComponentFixture, TestBed } from '@angular/core/testing';
...
+const testType: TestType = {
+  testTypeName: 'name',
+  testCode: '',
+  testNumber: '',
+  testExpiryDate: '',
+  testTypeStartTimestamp: '01/01/2001',
+  certificateNumber: ''
+}
+
+const fakeRecordMultipleTestTypes: TestResultModel = {
```
Comment (mcadenas-bjss) [inline src/app/features/test-record-summary/test-record-summary.component.spec.ts:25]: I recommend to use the `createMock` and `createMockList` functions to save space and consistency

File: src/app/features/test-record-summary/test-record-summary.component.ts [hunk 8]
```diff
@@ -11,4 +11,9 @@ export class TestRecordSummaryComponent {
   @Input() testRecords: TestResultModel[] = [];
 
   constructor() {}
+
+  getTestTypeName(testResult: TestResultModel) {
+    const testNames = testResult.testTypes.map(t => t.testTypeName).join(',');
```
Comment (mcadenas-bjss) [inline src/app/features/test-record-summary/test-record-summary.component.ts:16]: return expression here and save lines.

File: src/app/features/search/search.component.ts (hunk 6 above)
Comment (tom-bjss) [reply src/app/features/search/search.component.ts:26]: Thank you

File: src/app/features/test-record-summary/test-record-summary.component.ts (hunk 8 above)
Comment (tom-bjss) [reply src/app/features/test-record-summary/test-record-summary.component.ts:16]: Always in two minds about this one for readability but the function name is semantic enough that I feel alright about it

File: src/app/features/test-record-summary/test-record-summary.component.spec.ts (hunk 3 above)
Comment (andrewf-bjss) [reply src/app/features/test-record-summary/test-record-summary.component.spec.ts:10]: Thanks for the clarification @mcadenas-bjss - if we can keep the 2 model imports the same, the either path is fine  :)

File: src/app/features/search/search.component.ts [hunk 9]
```diff
@@ -19,11 +19,11 @@ export class SearchComponent {
 
   public searchTechRecords(searchTerm: string) {
     const searchErrorMessage = 'You must provide a vehicle registration mark, trailer ID or vehicle identification number.';
-
+    const trimmedSearchTerm = searchTerm.trim();
```
Comment (danielchildsbjss) [inline src/app/features/search/search.component.ts:22]: we could just do 
searchTerm = searchTerm.trim();
as we don't need the original, and would fix the bug below

**Review by andrewf-bjss — APPROVED** [review]

**Review by NiallBunting — APPROVED** [review]

**Review by aaron-downing — APPROVED** [review]

**Pull request merged**